Initial gRPC CLI

alesj opened the pull request:
Simple gcurl-like CLI -- currently supporting gRPC list, describe and invoke.

alesj committed:
Initial gRPC CLI

iocanel reviewed (approved):
PR looks good to me.

iocanel commented:
It's not clear to me why CI barfs.

alesj commented:
@gsmet any idea? ^

cescoffier commented on extensions/grpc/cli/src/main/java/io/quarkus/grpc/cli/GrpcCommand.java:
```diff
@@ -0,0 +1,27 @@
+import picocli.CommandLine.Spec;
+
+@TopCommand
+@Command(name = "grpc", sortOptions = false, header = "grpc CLI", subcommands = {
```
Why CLI in the header?

alesj replied:
So just dropping the CLI?

cescoffier commented on extensions/grpc/cli/src/main/java/io/quarkus/grpc/cli/GrpcCommand.java:
```diff
@@ -0,0 +1,27 @@
+    @Spec
+    protected CommandSpec spec;
+
+    @CommandLine.Option(names = { "-h", "--help" }, usageHelp = true, description = "Display this help message.")
```
Isn't this generated by picocli automatically?

alesj replied:
Hmmm, no idea -- copied this from one of the existing CLI commands ...
@iocanel ^ ?

alesj replied:
@iocanel can this be dropped? ^

iocanel replied:
I think that it can be possibly dropped.

cescoffier commented on extensions/grpc/cli/src/main/resources/application.properties:
```diff
@@ -0,0 +1,4 @@
+quarkus.log.level=INFO
+quarkus.banner.enabled=false
+quarkus.package.jar.type=uber-jar
+quarkus.package.jar.add-runner-suffix=false
```
This won't work I believe.
When building Quarkus from scratch, you won't find the plugin, because the plugin is built. 

In another case, I had to switch to the assembly maven plugin (would love to switch back to the Quarkus plugin).

alesj replied:
Hmm, yeah, this could be a problem ... since my copied example was from Quarkiverse, hence made it more easy.
Any ideas / suggestions what can be done?

cescoffier replied:
Yes :-) I'm going to push a PR today that does this.

maxandersen replied:
is there a reason we do uberjar here? why not just reuse users most likely already downloaded deps for this cli?

cescoffier replied:
@maxandersen seems like Uber-jar is the way to distribute self-contained commands. At least that's what I got from the doc and when I tried the "fast-jar" approach, it did not work (and anyway, fast jar would need quarkus before quarkus)

maxandersen replied:
yeah okey - if the actual cli is a quarkus app and not "just" a java app you need it. It would be interesting if we had a way to make a jar that has only the classes that need changing and rest could just be fetched using dependencies.

maxandersen replied:
hmm - even uber-jar would need quarkus before quarkus would it not?

cescoffier replied:
Using Quarkus for these commands works, but yes it might be overkill. 

By using the assembly plugin it works because it waits for all dependencies to be built before building that fat jar. Having a plugin dependency is slightly more complicated and fail in all my attempts from scratch. Funny, I had the same issue more than 10 years ago where I had to have profiles to build the plugin and then the modules using the plugin.

aloubyansky replied:
We should be able to build the Maven plugin right after core-deployment to make it available for other extensions.

cescoffier reviewed (dismissed):
We need to verify if it can be built from scratch.

Also, Some tests would be useful.

alesj commented:
> We need to verify if it can be built from scratch.

Scratch? Wdym?

> Also, Some tests would be useful.

Yeah ... is there some existing example of it?
So I don't re-discover water :-)

cescoffier commented:
Look at https://github.com/quarkusio/quarkus/pull/41418

alesj commented:
@cescoffier OK, added assembly + test(s)

gsmet commented:
@aloubyansky I'm not sure what's going on here but I have two entries:

```
{
    "name" : "SmallRye Context Propagation",
    "description" : "Propagate contexts between managed threads in reactive applications",
    "metadata" : { 
      "config" : [ "mp.context." ],
      "built-with-quarkus-core" : "999-SNAPSHOT",
      "guide" : "https://quarkus.io/guides/context-propagation",
      "short-name" : "context propagation",
      "scm-url" : "https://github.com/quarkusio/quarkus",
      "categories" : [ "reactive" ],
      "extension-dependencies" : [ "io.quarkus:quarkus-core", "io.quarkus:quarkus-arc" ],
      "keywords" : [ "smallrye-context-propagation", "microprofile-context-propagation", "context-propagation", "context", "reactive" ],
      "minimum-java-version" : "17",
      "requires-quarkus-core" : "[999.0,)",
      "status" : "stable"
    },
    "artifact" : "io.quarkus:quarkus-smallrye-context-propagation::jar:999-SNAPSHOT",
    "origins" : [ "io.quarkus:quarkus-bom-quarkus-platform-descriptor:999-SNAPSHOT:json:999-SNAPSHOT" ]
  }
```
in the devtools/bom-descriptor-json/target/quarkus-bom-quarkus-platform-descriptor-999-SNAPSHOT-999-SNAPSHOT.json file and it causes this build to fail.

I'm adding a `uniq` in the extension verification script but I wouldn't expect the same artifact to be twice in the file.

gsmet committed:
Avoid duplicates in update-extension-dependencies.sh

I don't know how we can end up with duplicates in the JSON descriptor
but let's make sure it doesn't affect our build.

aloubyansky commented:
Is it in this PR only @gsmet ?

gsmet commented:
@aloubyansky yes

aloubyansky commented:
Interesting, I'll check it out

gsmet commented:
Something is fishy but I can't figure out what.

gsmet commented:
Even the fact that the SR-CP deployment dependency is required for the IT is odd.

aloubyansky commented:
Simply building the repo the usual way does not result duplication. Running the script now.

aloubyansky commented:
It's clean. Can you reproduce it @gsmet ?

gsmet commented:
Well, I was reproducing it locally, yes. But not the first time I built it as I had a clean build once.

But today, I ended up with the duplication every time.

aloubyansky commented:
How exactly do you build the repo to reproduce it?

gsmet committed:
Avoid having an extension dependency in quarkus-grpc-api

It is not an extension so it should not depend on extensions.

gsmet commented:
I think we will need to take a big step back as to what is an extension or not in the gRPC stuff as we had numerous issues with that and, while we were able to work around them, some are still around.

I will merge it for 3.14 as I think it's an interesting addition and we can improve on it.
